# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/AudioData.cpp
# Commit: 1d7725b37869
# Full Hash: 1d7725b37869aa40919dfc4b3e0263ef35e1d7b9
# Author: Paul Adenot <paul@paul.cx>
# Date: 2024-07-08 16:33:59
# Description:
#   Bug 1902555 - Handle overflow in buffer size check in IsValidAudioData. r=chunmin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D215634
# ==============================================================================

diff -r 7c5450aa6e8e -r 1d7725b37869 dom/media/webcodecs/AudioData.cpp
--- a/dom/media/webcodecs/AudioData.cpp	Mon Jul 08 09:40:53 2024 +0000
+++ b/dom/media/webcodecs/AudioData.cpp	Mon Jul 08 10:29:06 2024 +0000
@@ -168,18 +168,27 @@
     return Err(msg);
   }
 
-  uint64_t totalSamples = aInit.mNumberOfFrames * aInit.mNumberOfChannels;
-  uint32_t bytesPerSamples = BytesPerSamples(aInit.mFormat);
-  uint64_t totalSize = totalSamples * bytesPerSamples;
+  CheckedInt<uint64_t> bytesNeeded = aInit.mNumberOfFrames;
+  bytesNeeded *= aInit.mNumberOfChannels;
+  bytesNeeded *= BytesPerSamples(aInit.mFormat);
+
+  if (!bytesNeeded.isValid()) {
+    auto msg = nsPrintfCString(
+        "Overflow when computing the number of bytes needed to hold audio "
+        "samples");
+    LOGD("%s", msg.get());
+    return Err(msg);
+  }
+
   uint64_t arraySizeBytes = ProcessTypedArraysFixed(
       aInit.mData, [&](const Span<uint8_t>& aData) -> uint64_t {
         return aData.LengthBytes();
       });
-  if (arraySizeBytes < totalSize) {
+  if (arraySizeBytes < bytesNeeded.value()) {
     auto msg =
         nsPrintfCString("Array of size %" PRIu64
                         " not big enough, should be at least %" PRIu64 " bytes",
-                        arraySizeBytes, totalSize);
+                        arraySizeBytes, bytesNeeded.value());
     LOGD("%s", msg.get());
     return Err(msg);
   }